e2e/tests/transfer/forwarding_test.go

```diff
@@ -108,3 +109,99 @@ func (s *TransferForwardingTestSuite) testForwardingThreeChains(lastChainVersion
+
+	channelAtoB := s.GetChainAChannel()
+	channelBtoC := s.GetChainChannel(testsuite.ChainChannelPair{ChainIdx: 1, ChannelIdx: 1})
+
+	chainAWallet := s.CreateUserOnChainA(ctx, testvalues.StartingTokenAmount)
+	chainAAddress := chainAWallet.FormattedAddress()
+	chainADenom := chainA.Config().Denom
+
+	chainBWallet := s.CreateUserOnChainB(ctx, testvalues.StartingTokenAmount)
+	chainBAddress := chainBWallet.FormattedAddress()
+
+	chainCWallet := s.CreateUserOnChainC(ctx, testvalues.StartingTokenAmount)
+	chainCAddress := chainCWallet.FormattedAddress()
+
+	t.Run("IBC transfer from A to B", func(t *testing.T) {
+		inFiveMinutes := time.Now().Add(5 * time.Minute).UnixNano()
+
+		msgTransfer := testsuite.GetMsgTransfer(
+			channelAtoB.PortID,
+			channelAtoB.ChannelID,
+			transfertypes.V2,
+			testvalues.DefaultTransferCoins(chainADenom),
+			chainAAddress,
+			chainBAddress,
+			clienttypes.ZeroHeight(),
+			uint64(inFiveMinutes),
+			"",
+			nil)
+		resp := s.BroadcastMessages(ctx, chainA, chainAWallet, msgTransfer)
+		s.AssertTxSuccess(resp)
+	})
+
+	t.Run("start relayer", func(t *testing.T) {
+		s.StartRelayer(relayer)
+	})
+
+	chainBDenom := transfertypes.NewDenom(chainADenom, transfertypes.NewHop(channelAtoB.Counterparty.PortID, channelAtoB.Counterparty.ChannelID))
+	t.Run("packet has reached B", func(t *testing.T) {
+		s.AssertPacketRelayed(ctx, chainA, channelAtoB.PortID, channelAtoB.ChannelID, 1)
+
+		balance, err := query.Balance(ctx, chainB, chainBAddress, chainBDenom.IBCDenom())
+		s.Require().NoError(err)
+
+		s.Require().Equal(testvalues.IBCTransferAmount, balance.Int64())
+	})
+
+	t.Run("IBC transfer from B (unwind) to C through A", func(t *testing.T) {
+		inFiveMinutes := time.Now().Add(5 * time.Minute).UnixNano()
+
+		forwarding := transfertypes.NewForwarding(
+			true,
+			transfertypes.NewHop(channelAtoB.PortID, channelAtoB.ChannelID),
+			transfertypes.NewHop(channelBtoC.PortID, channelBtoC.ChannelID),
+		)
+		msgTransfer := testsuite.GetMsgTransfer(
+			"",
+			"",
+			transfertypes.V2,
+			testvalues.DefaultTransferCoins(chainBDenom.IBCDenom()),
+			chainBAddress,
+			chainCAddress,
+			clienttypes.ZeroHeight(),
+			uint64(inFiveMinutes),
+			"",
+			forwarding)
+		resp := s.BroadcastMessages(ctx, chainB, chainBWallet, msgTransfer)
+		s.AssertTxSuccess(resp)
+	})
+	t.Run("packet has reached C", func(t *testing.T) {
+		chainCDenom := transfertypes.NewDenom(chainADenom,
+			transfertypes.NewHop(channelAtoB.Counterparty.PortID, channelAtoB.Counterparty.ChannelID),
+			transfertypes.NewHop(channelBtoC.Counterparty.PortID, channelBtoC.Counterparty.ChannelID),
+		)
+
+		err := test.WaitForCondition(time.Minute*10, time.Second*30, func() (bool, error) {
+			balance, err := query.Balance(ctx, chainC, chainCAddress, chainCDenom.IBCDenom())
+			if err != nil {
+				return false, err
+			}
+			return balance.Int64() == testvalues.IBCTransferAmount, nil
+		})
+		s.Require().NoError(err)
+	})
```

Comment:
Actually, it might be good, if possible, to check that the packet was acknowledge on chain B, so that we make sure that the relayer handles the asynchronous acknowledgements.

Reply:
I discussed with Damian and Cian ad we agreed that that check would be good to have but there is no straightforward way to add it right now and it would require more work. We added a check that the packet has been relayed.

Reply:
All right. Thanks, @bznein. Yeah, I suspected it was not straightforward because I didn't remember from the top of my head seeing it in any other e2e test... Would you like to open an issue so that we can track the improvement?

Reply:
Done! #6786